codegen/src/main/kotlin/software/amazon/smithy/rust/codegen/smithy/protocols/serialize/EventStreamMarshallerGenerator.kt

```diff
@@ -99,21 +106,59 @@ class EventStreamMarshallerGenerator(
         }
     }
 
-    private fun RustWriter.serializeEvent(struct: StructureShape) {
-        for (member in struct.members()) {
+    private fun RustWriter.renderMarshallEvent(unionMember: MemberShape, struct: StructureShape) {
+        val headerMembers = struct.members().filter { it.hasTrait<EventHeaderTrait>() }
+        val payloadMember = struct.members().firstOrNull { it.hasTrait<EventPayloadTrait>() }
+        for (member in headerMembers) {
             val memberName = symbolProvider.toMemberName(member)
             val target = model.expectShape(member.target)
-            if (member.hasTrait<EventPayloadTrait>()) {
-                serializeUnionMember(memberName, member, target)
-            } else if (member.hasTrait<EventHeaderTrait>()) {
-                TODO("TODO(EventStream): Implement @eventHeader trait")
-            } else {
-                throw IllegalStateException("Event Stream members must be a header or payload")
+            renderMarshallEventHeader(memberName, member, target)
+        }
+        if (payloadMember != null) {
+            val memberName = symbolProvider.toMemberName(payloadMember)
+            val target = model.expectShape(payloadMember.target)
+            renderMarshallEventPayload("inner.$memberName", payloadMember, target)
+        } else if (headerMembers.isEmpty()) {
+            renderMarshallEventPayload("inner", unionMember, struct)
+        } else {
+            addStringHeader(":content-type", "application/octet-stream".dq() + ".into()")
+            rust("Vec::new()")
+        }
```

Comment:
I don't really understand the three variants here & why only the last one sets a content type